File: packages/mobile-sdk/android/kintsugi/src/main/kotlin/com/kintsugi/KintsugiClient.kt
package com.kintsugi

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.*
import kotlinx.serialization.json.*
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.toRequestBody
import java.util.UUID
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

/**
 * Kintsugi Android SDK —— 对齐 @kintsugi/sdk / KintsugiKit.swift 的 Instant API 形状。
 *
 * 用法:
 *   val client = KintsugiClient(
 *       baseUrl = "https://api.kintsugi.example.com",
 *       appCode = "app-demo0001",
 *       auth = KintsugiAuth.Token("eyJ...")
 *   )
 *   val result = client.filter("dsxxxx", FilterRequest(pageSize = 20))
 */
sealed class KintsugiAuth {
    data object None : KintsugiAuth()
    data class Token(val token: String) : KintsugiAuth()
    data class AccessKey(val accessKey: String, val secretKey: String) : KintsugiAuth()
}

class KintsugiClient(
    val baseUrl: String,
    val appCode: String,
    val auth: KintsugiAuth = KintsugiAuth.None,
    val httpClient: OkHttpClient = OkHttpClient(),
    val json: Json = Json { ignoreUnknownKeys = true; encodeDefaults = false },
) {
    private val jsonMt = "application/json; charset=utf-8".toMediaType()

    suspend inline fun <reified T> filter(
        datasetCode: String,
        body: FilterRequest = FilterRequest(),
    ): FilterResult<T> =
        request("POST", "/api/apps/$appCode/ds/$datasetCode/filter", json.encodeToString(body))

    suspend inline fun <reified T> getOne(datasetCode: String, id: String): T =
        request("GET", "/api/apps/$appCode/ds/$datasetCode/$id", null)

    suspend inline fun <reified T, reified U> create(datasetCode: String, data: U): T =
        request("POST", "/api/apps/$appCode/ds/$datasetCode", json.encodeToString(data))

    suspend inline fun <reified T, reified U> update(datasetCode: String, id: String, data: U): T =
        request("PATCH", "/api/apps/$appCode/ds/$datasetCode/$id", json.encodeToString(data))

    suspend fun delete(datasetCode: String, id: String): DeleteResult =
        request("DELETE", "/api/apps/$appCode/ds/$datasetCode/$id", null)

    suspend fun askChats(question: String): ChatsAskResult {
        val body = json.encodeToString(ChatsAskRequest(appCode, question, null))
        return request("POST", "/api/chats/ask", body)
    }

    suspend inline fun <reified T> request(method: String, path: String, jsonBody: String?): T {
        val text = requestRaw(method, path, jsonBody)
        return json.decodeFromString(text)
    }

    suspend fun requestRaw(method: String, path: String, jsonBody: String?): String = withContext(Dispatchers.IO) {
        val url = baseUrl.trimEnd('/') + path
        val reqBuilder = Request.Builder().url(url)

        val body: String = jsonBody ?: ""
        when (method.uppercase()) {
            "GET" -> reqBuilder.get()
            "DELETE" -> reqBuilder.delete(body.toRequestBody(jsonMt))
            else -> reqBuilder.method(method.uppercase(), body.toRequestBody(jsonMt))
        }

        when (auth) {
            is KintsugiAuth.None -> Unit
            is KintsugiAuth.Token -> reqBuilder.header("Authorization", "Bearer ${auth.token}")
            is KintsugiAuth.AccessKey -> {
                val ts = (System.currentTimeMillis() / 1000).toString()
                val nonce = UUID.randomUUID().toString().replace("-", "").substring(0, 16).lowercase()
                val canonical = "${method.uppercase()}\n$path\n$ts\n$nonce\n$body"
                val sig = hmacSha256(auth.secretKey, canonical)
                reqBuilder.header("X-Access-Key", auth.accessKey)
                    .header("X-Timestamp", ts)
                    .header("X-Nonce", nonce)
                    .header("X-Signature", sig)
            }
        }

        httpClient.newCall(reqBuilder.build()).execute().use { res ->
            val txt = res.body?.string() ?: ""
            if (!res.isSuccessful) throw KintsugiHttpException(res.code, txt)
            txt
        }
    }

    private fun hmacSha256(key: String, message: String): String {
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(key.toByteArray(), "HmacSHA256"))
        return mac.doFinal(message.toByteArray()).joinToString("") { "%02x".format(it) }
    }
}

class KintsugiHttpException(val status: Int, val body: String) :
    RuntimeException("HTTP $status: ${body.take(200)}")

// ---- DTOs ----
@Serializable
data class FilterClause(val field: String, val op: String, val value: JsonElement? = null)

@Serializable
data class SortOrder(val field: String, val direction: String? = null)

@Serializable
data class FilterRequest(
    @SerialName("where") val wheres: List<FilterClause>? = null,
    val orderBy: List<SortOrder>? = null,
    val page: Int? = null,
    val pageSize: Int? = null,
    val select: List<String>? = null,
    val includeDeleted: Boolean? = null,
)

@Serializable
data class FilterResult<T>(
    val data: List<T>,
    val total: Int,
    val page: Int,
    val pageSize: Int,
)

@Serializable
data class DeleteResult(val ok: Boolean, val softDeleted: Boolean)

@Serializable
data class ChatsAskRequest(val appCode: String, val question: String, val maxTables: Int?)

@Serializable
data class ChatsAskResult(
    val sql: String,
    val explanation: String,
    val rowCount: Int,
)

File: packages/mobile-sdk/ios/Sources/KintsugiKit/KintsugiClient.swift
import Foundation
import CryptoKit

/// Kintsugi iOS SDK —— 对齐 @kintsugi/sdk 的 Instant API 形状。
///
/// 用法：
///   let client = KintsugiClient(
///       baseURL: URL(string: "https://api.kintsugi.example.com")!,
///       appCode: "app-demo0001",
///       auth: .token("eyJ...")
///   )
///   let result: FilterResult<Goods> = try await client.filter(datasetCode: "dsxxx", body: .init(pageSize: 20))

public enum KintsugiAuth {
    case none
    case token(String)
    /// HMAC-SHA256 签名模式（对应服务端 AccessKey）
    case accessKey(key: String, secret: String)
}

public struct KintsugiClient {
    public let baseURL: URL
    public let appCode: String
    public var auth: KintsugiAuth
    public var session: URLSession

    public init(baseURL: URL, appCode: String, auth: KintsugiAuth = .none, session: URLSession = .shared) {
        self.baseURL = baseURL
        self.appCode = appCode
        self.auth = auth
        self.session = session
    }

    // MARK: - Instant API

    public func filter<T: Decodable>(datasetCode: String, body: FilterRequest) async throws -> FilterResult<T> {
        try await request("POST", path: "/api/apps/\(appCode)/ds/\(datasetCode)/filter", body: body)
    }

    public func getOne<T: Decodable>(datasetCode: String, id: String) async throws -> T {
        try await request("GET", path: "/api/apps/\(appCode)/ds/\(datasetCode)/\(id)", body: Empty())
    }

    public func create<In: Encodable, Out: Decodable>(datasetCode: String, data: In) async throws -> Out {
        try await request("POST", path: "/api/apps/\(appCode)/ds/\(datasetCode)", body: data)
    }

    public func update<In: Encodable, Out: Decodable>(datasetCode: String, id: String, data: In) async throws -> Out {
        try await request("PATCH", path: "/api/apps/\(appCode)/ds/\(datasetCode)/\(id)", body: data)
    }

    public func delete(datasetCode: String, id: String) async throws -> DeleteResult {
        try await request("DELETE", path: "/api/apps/\(appCode)/ds/\(datasetCode)/\(id)", body: Empty())
    }

    // MARK: - Chats

    public func askChats(question: String) async throws -> ChatsAskResult {
        try await request("POST", path: "/api/chats/ask", body: ChatsAskRequest(appCode: appCode, question: question, maxTables: nil))
    }

    // MARK: - Core

    public func request<In: Encodable, Out: Decodable>(_ method: String, path: String, body: In) async throws -> Out {
        var req = URLRequest(url: baseURL.appendingPathComponent(path))
        req.httpMethod = method
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.setValue("application/json", forHTTPHeaderField: "Accept")

        var bodyData: Data = Data()
        if method != "GET" && !(body is Empty) {
            bodyData = try JSONEncoder().encode(body)
            req.httpBody = bodyData
        }

        switch auth {
        case .none: break
        case .token(let t):
            req.setValue("Bearer \(t)", forHTTPHeaderField: "Authorization")
        case .accessKey(let key, let secret):
            let ts = String(Int(Date().timeIntervalSince1970))
            let nonce = UUID().uuidString.replacingOccurrences(of: "-", with: "").prefix(16).lowercased()
            let bodyStr = String(data: bodyData, encoding: .utf8) ?? ""
            let canonical = "\(method.uppercased())\n\(path)\n\(ts)\n\(nonce)\n\(bodyStr)"
            let sig = hmacSHA256(key: secret, message: canonical)
            req.setValue(key, forHTTPHeaderField: "X-Access-Key")
            req.setValue(ts, forHTTPHeaderField: "X-Timestamp")
            req.setValue(String(nonce), forHTTPHeaderField: "X-Nonce")
            req.setValue(sig, forHTTPHeaderField: "X-Signature")
        }

        let (data, res) = try await session.data(for: req)
        guard let http = res as? HTTPURLResponse else {
            throw KintsugiError.network("no http response")
        }
        guard (200..<300).contains(http.statusCode) else {
            throw KintsugiError.http(status: http.statusCode, body: String(data: data, encoding: .utf8) ?? "")
        }
        return try JSONDecoder().decode(Out.self, from: data)
    }
}

private func hmacSHA256(key: String, message: String) -> String {
    let keyData = Data(key.utf8)
    let msgData = Data(message.utf8)
    let sig = HMAC<SHA256>.authenticationCode(for: msgData, using: SymmetricKey(data: keyData))
    return sig.map { String(format: "%02x", $0) }.joined()
}

public enum KintsugiError: Error {
    case network(String)
    case http(status: Int, body: String)
}
